hs2-types/src/Hs2/Types/Type.hs

```diff
@@ -0,0 +1,24 @@
+module Hs2.Types.Type where
+
+type Cxt = Type
+
+data PromotableType
+  = ConTy Name
+  | VarTy Name
+  | TupleTy Int
+
+data Type
+  = ForallTy [Name] Type
+  | ConstrainTy Cxt Type
+  | AppTy Type Type
+  | RegularTy PromotableType
+  | PromotedTy PromotableType
+  | LitTy TyLit
+  | ArrowTy Type Type
```

Comment:
We can further desugar `ArrowTy x y` down to `AppTy (AppTy (ConTy "->") x) y`, and use a pattern to get this more usable.

Reply:
Also for `TupleTy`